### PR TITLE
Graceful shutdown: will fix data loss/duplicate (corner cases)

### PR DESCRIPTION
**Pull request by abenbachir**
Reduce OMS shutdown to 1 second.
Set 3 seconds timeout while root_agent is calling shutdown on each input plugin.
Send SIGTERM instead of SIGKILL to give the agent time to cleanup during shutdown.

### CODE REVIEW

File: build/buildRuby.sh [hunk 1]
```diff
@@ -287,6 +281,8 @@ echo "========================= Performing Building FluentD"
 cd ${FLUENTD_DIR}
 elevate bundle install --local
 elevate bundle exec rake build
+# currently 7 tests are failing
```
Comment (simathih) [inline build/buildRuby.sh]: is this related?

File: source/ext/patches/fluentd/in_exec.patch [hunk 2]
```diff
@@ -1,10 +1,59 @@
...
++
++        begin
++          # Need to reap the process because it will become a zombie process
++          Process.detach(@pid)
++        rescue #Errno::ECHILD, Errno::ESRCH, Errno::EPERM
++        end
++        # cleanup the thread
++        Thread.kill(@thread)
++        @thread.join(0.1)
+       end
```
Comment (simathih) [inline source/ext/patches/fluentd/in_exec.patch]: do we really care about the data lost here? how many CRIs do we see due to this?

Comment (simathih) [reply source/ext/patches/fluentd/in_exec.patch]: my worry is this is deep in fluentd engine and could cause other unintended issues if not verified properly

Comment (abenbachir) [reply source/ext/patches/fluentd/in_exec.patch]: This change will only impact the shutdown sequence, if there is a crash that happen because of this change, it will not cause any issue since it will happens only during shudown.

**Comment by abenbachir** [comment]
This is a graceful shutdown, where fluentd will do:
- Clean-up.
- Flush memory to buffer.
- Release sub-process to prevent zombie process.

I have an upcoming PR that needs graceful shutdown.

File: build/buildRuby.sh (hunk 1 above)
Comment (abenbachir) [reply build/buildRuby.sh]: Not related, I wanted to enable fluentd unittests, but there are 7 tests already failing, it doesn't have to do with this change.

**Comment by abenbachir** [comment]
@hestolz with this PR, you can use the shutdown function to push to Geneva some QoS/errors.

File: source/code/plugins/in_oms_heartbeat.rb [hunk 3]
```diff
@@ -32,6 +32,8 @@ def start
     def shutdown
       if @interval
         @finished = true
+        log.debug("in_oms_hearthbeat: thread state before shutdown '#{@thread.status}'")
```
Comment (hestolz) [inline source/code/plugins/in_oms_heartbeat.rb]: `@log.debug`

File: source/code/plugins/in_dsc_monitor.rb [hunk 4]
```diff
@@ -102,7 +102,11 @@ def shutdown
       super
       @finished_check_install = true
       @finished_check_status = true
+      log.debug("in_dsc_monitor: thread 'check_install' state before shutdown was '#{@thread_check_install.status}'")
```
Comment (hestolz) [inline source/code/plugins/in_dsc_monitor.rb]: `@log.debug` for both

File: source/code/plugins/in_agent_telemetry.rb [hunk 5]
```diff
@@ -57,6 +57,8 @@ def start
     def shutdown
       if defined?(OMS::Configuration.telemetry_interval)
         @finished = true
+        log.debug("in_agent_telemetry: thread state before shutdown '#{@thread.status}'")
```
Comment (hestolz) [inline source/code/plugins/in_agent_telemetry.rb]: `@log.debug`

File: source/code/plugins/in_agent_telemetry.rb [hunk 6]
```diff
@@ -57,6 +57,8 @@ def start
...
+        @thread.exit
```
Comment (hestolz) [inline source/code/plugins/in_agent_telemetry.rb]: already added in my finally-merged qos-errors pr :)

Comment (abenbachir) [reply source/code/plugins/in_agent_telemetry.rb]: Great then I will remove this. :+1:
